cmd/clusterctl/client/alpha/rollout_restarter.go

```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package alpha
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/types"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/client/cluster"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/internal/util"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var validResourceTypes = []string{"machinedeployment"}
+
+func (r *rollout) ObjectRestarter(proxy cluster.Proxy, tuple util.ResourceTuple, namespace string) error {
+	switch tuple.Resource {
+	case "machinedeployment":
+		mdObj := &clusterv1.MachineDeployment{}
+		if err := getMachineDeployment(proxy, tuple.Name, namespace, mdObj); err != nil {
+			return errors.Wrapf(err, "failed to fetch %v/%v", tuple.Resource, tuple.Name)
+		}
+		if mdObj.Spec.Paused {
+			fmt.Printf("can't restart paused machinedeployment (run rollout resume first): %v/%v\n", tuple.Resource, tuple.Name)
+			return nil
+		}
+		if err := setRestartedAtAnnotation(proxy, tuple.Name, namespace); err != nil {
+			return err
+		}
+	case "kubeadmcontrolplane":
+		return fmt.Errorf("restarting is not supported")
+	default:
+		return errors.Errorf("Invalid resource type %v. Valid values: %v", tuple.Resource, validResourceTypes)
+	}
+	return nil
+}
+
+// getMachineDeployment retrieves the MachineDeployment object corresponding to the name and namespace specified.
+func getMachineDeployment(proxy cluster.Proxy, name, namespace string, mdObj *clusterv1.MachineDeployment) error {
+	c, err := proxy.NewClient()
+	if err != nil {
+		return err
+	}
+	mdObjKey := client.ObjectKey{
+		Namespace: namespace,
+		Name:      name,
+	}
+	if err := c.Get(context.TODO(), mdObjKey, mdObj); err != nil {
+		return errors.Wrapf(err, "error reading %q %s/%s",
+			mdObj.GroupVersionKind(), mdObj.GetNamespace(), mdObj.GetName())
+	}
+	return nil
+}
+
+// setRestartedAtAnnotation sets the restartedAt annotation in the MachineDeployment's spec.template.objectmeta.
+func setRestartedAtAnnotation(proxy cluster.Proxy, name, namespace string) error {
+	patch := client.RawPatch(types.MergePatchType, []byte(fmt.Sprintf("{\"spec\":{\"template\":{\"metadata\":{\"annotations\":{\"cluster.x-k8s.io/restartedAt\":\"%v\"}}}}}", time.Now().Format(time.RFC3339))))
```

Comment:
Suggestion: We've been using `util/patch` in cluster-api to patch objects. See usage example in `controllers/machinedeployment_controller.go`.

Reply:
Ahh...ok. Didn't notice that.

Reply:
Regarding this line, could we have a test to assert that the expected annotation on the template of the machine deployment is set?

Reply:
Yes, I added it: Test_ObjectRestarter